File: interpreter/ilox/lox/lox.go
package lox

import (
	"bufio"
	"fmt"
	"ilox/lox/loxerr"
	"ilox/lox/scan"
	"ilox/lox/token"
	"os"
	"strings"
)

type Lox struct {
	err loxerr.Reporter
}

func New() Lox {
	return Lox{}
}

func usage() string {
	s := "Usage:\n"
	s += "  ilox                    runs REPL\n"
	s += "  ilox run [script]       runs the script\n"
	s += "  ilox format [script]    formats the script\n"
	s += "  ilox help\n"
	return s
}

func (l *Lox) Run(args []string) {
	switch len(args) {
	case 0:
		l.runPrompt()
		return
	case 1:
		switch args[0] {
		case "help":
			fmt.Println(usage())
			return
		}
	case 2:
		switch args[0] {
		case "run":
			l.runFile(args[1])
			return
		case "format":
			l.runFormatter(args[1])
			return
		}
	}

	fmt.Fprintf(os.Stderr, "Unrecognized arguments: %q\n", strings.Join(args, " "))
	fmt.Fprintf(os.Stderr, usage())
	os.Exit(64)
}

func (l *Lox) runPrompt() {
	l.err.UseStdout = true
	fmt.Printf("> ")

	input := bufio.NewScanner(os.Stdin)
	for input.Scan() {
		line := input.Text()
		if line == "exit" {
			return
		}

		l.run(line)
		l.err.Reset()
		fmt.Printf("> ")
	}
}

func (l *Lox) runFile(path string) {
	bytes, err := os.ReadFile(path)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(65)
	}

	l.run(string(bytes))
	if l.err.HasError() {
		os.Exit(65)
	}
}

func (l *Lox) runFormatter(path string) {
	bytes, err := os.ReadFile(path)
	if err != nil {
		fmt.Fprintln(os.Stderr, err)
		os.Exit(65)
	}

	scanner := scan.NewScanner(string(bytes), &l.err)
	tokens := scanner.ScanTokens()
	if l.err.HasError() {
		os.Exit(65)
	}

	src := token.Format(tokens)
	fmt.Println(src)
}

func (l *Lox) run(src string) {
	scanner := scan.NewScanner(src, &l.err)
	tokens := scanner.ScanTokens()
	for _, token := range tokens {
		fmt.Println(token)
	}
}

File: interpreter/swift/lox/Sources/lox/Lox.swift
import Foundation

@main
struct Lox {
    static var hadError = false
    
    static func main() {
        do {
            try runFromArgs()
        } catch {
            print(error, to: &standardError)
            exit(64)
        }
    }
    
    private static func runFromArgs() throws {
        switch Args.get() {
        case let .unrecognized(args):
            print("error: unrecognized arguments: \(args)\n", to: &standardError)
            print(Args.usage(), to: &standardError)
            exit(64)
        case .help:
            print(Args.usage())
        case .prompt:
            try runPrompt()
        case let .run(path):
            try runFile(path: path)
        case .format(_):
            print("error: format is not yet implemented", to: &standardError)
            exit(64)
        }
    }
    
    private static func runFile(path: String) throws {
        let contents = try String(contentsOfFile: path, encoding: .utf8)
        try run(source: contents)
        if hadError {
            exit(65)
        }
    }
    
    private static func runPrompt() throws {
        let printPrompt = { print("> ", terminator: "") }
        printPrompt()
        while let line = readLine() {
            try run(source: line)
            hadError = false
            printPrompt()
        }
    }
    
    private static func run(source: String) throws {
        let scanner = Scanner(source: source)
        let tokens = scanner.scanTokens()
        for token in tokens {
            print(token)
        }
    }
    
    // TODO: Consider refactoring
    static func error(line: Int, message: String) {
        report(line: line, at: "", message: message)
    }
    
    static func report(line: Int, at: String, message: String) {
        print("[line \(line)] Error\(at): \(message)", to: &standardError)
        hadError = true
    }
}
